grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -13,6 +13,40 @@ class PrepaymentInvoiceService {
 
     InvoiceService invoiceService
 
+    Invoice generatePrepaymentInvoice(Order order) {
+        if (order.orderItems.any { it.hasInvoices } || order.orderAdjustments.any { it.hasInvoices }) {
+            throw new Exception("Some order items or order adjustments for this order already have been invoiced")
+        }
+
+        Invoice invoice = invoiceService.createFromOrder(order)
+        invoice.invoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+        invoiceService.createOrUpdateVendorInvoiceNumber(invoice, order.orderNumber + Constants.PREPAYMENT_INVOICE_SUFFIX)
+
+        if (order?.orderItems?.empty && order?.orderAdjustments?.empty) {
+            throw new Exception("No order items or order adjustments found for given order")
+        }
+
+        BigDecimal prepaymentPercent = order.paymentTerm?.prepaymentPercent?:100
```

Comment:
This 100 should probably be in a `DEFAULT_PAYMENT_PERCENT` constant?

Actually (and I know this probably is pre-existing code), now that I think about it, is this even safe to do? These are the payment terms we have:

```
| id    | version | name                 | code  | description | date_created        | last_updated        | prepayment_percent | days_to_payment |
|-------|---------|----------------------|-------|-------------|---------------------|---------------------|--------------------|-----------------|
| COD   | 0       | Cash On Delivery     | COD   | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | NULL               | NULL            |
| NET30 | 0       | Net 30               | NET30 | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | NULL               | 30              |
| PIA   | 0       | Payment In Advance   | PIA   | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | NULL               | NULL            |
| PP100 | 0       | Prepayment 100       | PP100 | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | 100                | NULL            |
| PP25  | 0       | Prepayment 25/Net 30 | PP25  | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | 25                 | 30              |
| PP50  | 0       | Prepayment 50/Net 30 | PP50  | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | 50                 | 30              |
| US    | 0       | Upon shipment        | US    | NULL        | 2023-08-14 15:02:59 | 2023-08-14 15:02:59 | NULL               | 30              |

```

So I'm hoping we're not allowing the user to use non-prepayment terms here. Not sure how PIA differs from PP100 but I'm assuming that's why we need this null check.

Anyways, this isn't new code so I'm not asking for a refactor, just noting it down as it seemed a bit strange.

Reply:
Added the constant, but yeah this was an old calculation, I think it was done like that on purpose - if a term does not have a prepayment percentage we always prepay full amount